pkg/summarizer/flakiness.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package summarizer
+
+import (
+	"context"
+	"regexp"
+
+	"github.com/GoogleCloudPlatform/testgrid/internal/result"
+	"github.com/GoogleCloudPlatform/testgrid/pb/state"
+	summarypb "github.com/GoogleCloudPlatform/testgrid/pb/summary"
+	"github.com/GoogleCloudPlatform/testgrid/util/types"
+)
+
+const (
+	// Keep in mind that flakiness is measured as out of 100, i.e. 23 not .23
+	defaultFlakiness = 50.0
+	minRuns          = 0
+)
+
+var (
+	infraRegex = regexp.MustCompile(`^\w+$`)
+)
+
+type flakinessAnalyzer interface {
+	GetFlakiness(results []*types.Result, minRuns int, startDate int, endDate int, tab string) *summarypb.HealthinessInfo
+}
+
+// CalculateHealthiness extracts the test run data from each row (which represents a test)
+// of the Grid and then analyzes it with an implementation of flakinessAnalyzer, which has
+// implementations in the subdir naive and can be injected as needed.
+func CalculateHealthiness(grid *state.Grid, analyzer flakinessAnalyzer, startTime int, endTime int, tab string) *summarypb.HealthinessInfo {
+	results := parseGrid(grid, startTime, endTime)
+
+	return analyzer.GetFlakiness(results, minRuns, startTime, endTime, tab)
+}
+
+func parseGrid(grid *state.Grid, startTime int, endTime int) []*types.Result {
+	// Get the relevant data for flakiness from each Grid (which represents
+	// a dashboard tab) as a list of Result structs
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// We create maps because result.Map returns a map where we can access each result
+	// through the test name, and at each instance we can increment our types.Result
+	// using the same key. At the end we can filter out those types.Result that had
+	// 0 of all counts.
+	resultsMap := make(map[string]*types.Result, 0)
+	gridRows := make(map[string]*state.Row)
+	for i, row := range grid.Rows {
+		gridRows[row.Name] = grid.Rows[i]
+	}
+	// result.Map is written in a way that assumes each test/row name is unique
+	rowResults := result.Map(ctx, grid.Rows)
+
+	for i := 0; i < len(grid.Columns); i++ {
+		if !isWithinTimeFrame(grid.Columns[i], startTime, endTime) {
+			continue
+		}
+
+		for key, ch := range rowResults {
+			if resultsMap[key] == nil {
```

Comment:
use 2nd return value to check key existence, and let's take key in the the NewResult():

```
result, ok := resultsMap[key]; !ok {
    resultsMap[key] = types.NewResult(key)
}
```

Reply:
Done